engine/src/flutter/lib/ui/painting/multi_frame_codec.cc

```diff
@@ -144,10 +151,31 @@ MultiFrameCodec::State::GetNextFrameImage(
 
 #if IMPELLER_SUPPORTS_RENDERING
   if (is_impeller_enabled_) {
-    // This is safe regardless of whether the GPU is available or not because
-    // without mipmap creation there is no command buffer encoding done.
-    return ImageDecoderImpeller::UploadTextureToStorage(
-        impeller_context, std::make_shared<SkBitmap>(bitmap));
+    std::shared_ptr<impeller::DeviceBuffer> device_buffer =
+        impeller_context->GetResourceAllocator()->CreateBufferWithCopy(
+            fml::NonOwnedMapping(
+                reinterpret_cast<uint8_t*>(bitmap.getAddr(0, 0)),
+                bitmap.computeByteSize()));
+    if (device_buffer == nullptr) {
+      return std::make_pair(nullptr,
+                            "Failed to allocate device buffer for upload");
+    }
+
+    sk_sp<DlImage> image;
+    std::string error;
+    ImageDecoderImpeller::UploadTextureToPrivate(
+        [&image, &error](sk_sp<DlImage> image_result,
+                         std::string error_result) {
+          image = std::move(image_result);
+          error = std::move(error_result);
+        },
+        impeller_context,        //
+        device_buffer,           //
+        info,                    //
+        std::nullopt,            //
```

Comment:
Can you use parameter comments?  I'm not sure what `std::nullopt` is or the lambda above.

Reply:
Done